uvicorn/supervisors/multiprocess.py

```diff
@@ -32,39 +109,125 @@ def __init__(
+    def restart_all(self) -> None:
+        for idx, process in enumerate(tuple(self.processes)):
+            process.terminate()
+            process.join()
+            del self.processes[idx]
+            process = Process(self.config, self.target, self.sockets)
+            process.start()
+            self.processes.append(process)
+
+    def run(self) -> None:
+        message = f"Started parent process [{os.getpid()}]"
+        color_message = "Started parent process [{}]".format(click.style(str(os.getpid()), fg="cyan", bold=True))
+        logger.info(message, extra={"color_message": color_message})
+
+        self.init_processes()
+
+        while not self.should_exit.wait(0.5):
+            self.handle_signals()
+            self.keep_subprocess_alive()
+
+        self.terminate_all()
+        self.join_all()
+
+        message = f"Stopping parent process [{os.getpid()}]"
+        color_message = "Stopping parent process [{}]".format(click.style(str(os.getpid()), fg="cyan", bold=True))
         logger.info(message, extra={"color_message": color_message})
+
+    def keep_subprocess_alive(self) -> None:
+        for idx, process in enumerate(tuple(self.processes)):
+            if process.is_alive():
+                continue
+
+            process.kill()  # process is hung, kill it
+            process.join()
+            logger.info(f"Child process [{process.pid}] died")
+            del self.processes[idx]
+            process = Process(self.config, self.target, self.sockets)
+            process.start()
+            self.processes.append(process)
+
+    def handle_signals(self) -> None:
+        for sig in tuple(self.signal_queue):
+            self.signal_queue.remove(sig)
+            sig_name = UNIX_SIGNALS[sig]
+            sig_handler = getattr(self, f"handle_{sig_name.lower()}", None)
+            if sig_handler is not None:
+                sig_handler()
+            else:
+                logger.info(f"Received signal [{sig_name}], but nothing to do")
+
+    def handle_int(self) -> None:
+        if not self.should_exit.is_set():
+            self.should_exit.set()
+        else:
+            self.terminate_all()
+
+    def handle_term(self) -> None:
+        logger.info("Received SIGTERM, exiting")
+        if not self.should_exit.is_set():
+            self.should_exit.set()
+        else:
+            self.terminate_all()
+
+    def handle_break(self) -> None:
+        logger.info("Received SIGBREAK, exiting")
+        if not self.should_exit.is_set():
+            self.should_exit.set()
+        else:
+            self.terminate_all()
+
+    def handle_hup(self) -> None:
+        logger.info("Received SIGHUP, restarting processes")
+        self.restart_all()
+
+    def handle_ttin(self) -> None:
+        logger.info("Received SIGTTIN, increasing processes")
+        self.processes_num += 1
+        process = Process(self.config, self.target, self.sockets)
+        process.start()
+        self.processes.append(process)
```

Comment:
This pattern of `Process` instantiation, `start`, and `append` is scattered in several locations. This will IMO become an issue over time as features are built on top of this. I'd suggest centralizing that somehow in a nicer pattern that allows more flexibility.

Reply:
Excessive abstraction can cause more problems. I don't think a unified approach is needed until many of the new features you talk about are added. At least in my opinion, multi-process managers only have these functions at most (see gunicorn)